### ISSUE TITLE
[false-positive] Use ::class instead - low-case only strings should not be processed

### ISSUE THREAD

**Issue by PatrickBauer** [kalessil/phpinspectionsea#81]
Hi, 
unfortunately I can't post any more good documentation because I don't know what exactly happens, but the plugin suggests this fix for the following (simple) code:

```
public function _asText($value, $format) {
    return [
      'value' => $value,
      'format' => $format
    ];
  }
```

The 'value' key is the code part which is suggested, but it doesn't make any sense there.
Happy to answer any question about this code.

**Comment by kalessil** [comment]
Thank you for reporting @PatrickBauer, 2 thing would help:

* list of all imports in the file;
* result of quick-fix for the case (so we know exact class the inspection suggests);

**Comment by PatrickBauer** [comment]
Sure, this are the imports:

```
namespace Drupal\vs_imports\Transformer;

use Drupal\Component\Datetime\DateTimePlus;
use Drupal\taxonomy\Entity\Term;
use Drupal\vs_imports\Exception\ImportException;
use Drupal\vs_imports\Reader\ReaderInterface;
```

this is the quick fix:

```
  public function _asText($value, $format) {
    return [
      \value::class => $value,
      'format'      => $format
    ];
  }
```

**Comment by kalessil** [comment]
Okay, there is a `class value {...}` somewhere in the project, imports are not declaring it.

In this case you have to suppress the warning (expression level).
An alternative is to rename ` value` class into `Value`, but I'm not sure if you are able to do this.

**Comment by kalessil** [comment]
@PatrickBauer : would you please share some feedback here?

**Comment by PatrickBauer** [comment]
Yes, absolutely:
You're right, its a Drupal 8 Project which has a Value Class here: https://api.drupal.org/api/drupal/core!lib!Drupal!Core!Render!Element!Value.php/class/Value/8.2.x  (there is a "View source" button). 

I know I could suppress the warning, but its happening all over the source code (as 'value' is a really common name) and everyone using Drupal 8 will encounter these issues.

**Comment by PatrickBauer** [comment]
Actually there is another Value class from Doctrine: Doctrine\Common\Collections\Expr\Value

**Comment by kalessil** [comment]
Hm, then probably I should not inspect strings in lower case only. Let's give it a try with the next release.

**Comment by PatrickBauer** [comment]
Isn't it possible to match only to available classes? In my case the Value classes are not in the current namespace and are not imported via "use". That means without using the fully-qualified name, I shouldn't even be able to use these classes. In this case the provided quick-fix produces an invalid result.

**Comment by PatrickBauer** [comment]
Interesting, as you can see here:
https://www.dropbox.com/s/ahc17fbfq0fqhko/Screenshot%202016-12-20%2010.38.40.png?dl=0

PHPStorm seems to find test cases from php-parser. The file itself looks like this:
https://www.dropbox.com/s/nqousstaihvfj2h/Screenshot%202016-12-20%2010.40.01.png?dl=0

Maybe this helps in this case?

**Comment by kalessil** [comment]
Classes are resolved project-wide, not only current namespace (because of global classes).
That's why QF imposed a global class, that's by design)

Yep, that's the source of the problem, the classes seems to be defined in the root NS and the inspection recognizes it.

What file type is on the 2nd screenshot?

**Comment by PatrickBauer** [comment]
Both Value classes from before are not defined in the root namespace but under Drupal\Core\Render\Element and under Doctrine\Common\Collections\Expr. 

The file type from the second screenshot is nothing I recognize. The file extension is ".test". It seems to be an abstract syntax tree, see https://github.com/nikic/PHP-Parser. I don't know why PHPStorm finds these test cases as classes. The php-parser itself seems to be a core dependency of Drupal 8.

**Comment by kalessil** [comment]
Okay, I need to look into the issue based on this information. 
Most probably nothing is going to happen before January.

**Comment by PatrickBauer** [comment]
Sure, no problem. I'm just gonna disable the inspection for now.

**Comment by kalessil** [comment]
This would be the best option until a new release is out.

Btw, I'm curious about couple thing:
- how long you have been using the plugin and how did you run into it?
- was it many inspections which annoyed you, so you have to disable them?

**Comment by PatrickBauer** [comment]
Sure, happy to answer these questions. 

I've used the plugin for several months now. I've found out about it in one of the reddit.com/r/php threads and really liked it. It helps me to find a lot of small things that would normally slip through  (like in_array's third parameter). 

The only inspection I might disable is the long inheritance chain, but the inspection itself is great and would normally help. Just in my case I don't want to rewrite it ;)

**Comment by kalessil** [comment]
Great ;)

That's a good feedback: seems most of the sharp corners already worked out in the default configuration.

Based on your feedback I'll add online documentation for the disconnected statements inspection in the next release, thank you for your points.

**Comment by b1rdex** [comment]
I think `value` is treated as class because of `TRAIT` word before it. (Assumed from this screenshot: https://www.dropbox.com/s/nqousstaihvfj2h/Screenshot%202016-12-20%2010.40.01.png?dl=0 )

Here is that file: https://github.com/nikic/PHP-Parser/blob/master/test/code/parser/semiReserved.test
It's starts with `<?php` so PHPStorm tryes to read it as php.

**Comment by b1rdex** [comment]
My PHPStorm doesn't treat that `.test` file as php code.
@PatrickBauer don't you registered `.test` extensions to be treated as php files in PHPStorm? Settrings → Editor → File types

**Comment by b1rdex** [comment]
As for issue: I vote for igroring strings in lower case only. PSR-1/2/12 states that class names must be in Studly/PascalCase.

**Comment by bram123** [comment]
Or you can ignore all classes in the global namespace.

**Comment by b1rdex** [comment]
@bram123 no, that's not acceptable for legacy projects.

**Comment by kalessil** [comment]
We have to ignore lower-case only strings, as this also spotted by a performance profiler.

[This fix](https://github.com/kalessil/phpinspectionsea/commit/cbc35c7697875cb46c3cd209c85313488b408111#diff-929dbc0001ec4885788e7e47afe9f0e7R94) covers the issue and brings 25% performance improvement for the inspection (the inspection was one of the most CPU greedy).

**Comment by kalessil** [comment]
Overall performance looking good, the case seems to be covered.
